### PR TITLE
add expiration_time to cache

### PR DESCRIPTION
**Pull request by vera1118**
Hi, 
I want to use your great package, but I need a feature to deal with expiration for cache elements. So I added expiration_time to alru_cache and wrote a simple test. Do you want to update the cache? Let me know if you want and feel free to give suggestions. Thank you!

### CODE REVIEW

**Comment by hellysmile** [comment]
Hey, it mostly should be separate package cuz lru is not designed for that.

My suggestion is just each N seconds clear lru at all, if You need some kind of ttl

Also I've seen https://github.com/krkd/aiottl , which looks promising for ttl cache, but it is seems not fully ready yet

File: async_lru.py [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import asyncio
 from collections import OrderedDict
 from functools import _CacheInfo, _make_key, partial, wraps
+from time import time
```
Comment (hellysmile) [inline async_lru.py]: time.time is danger, take a look on https://docs.python.org/3/library/time.html#time.monotonic or loop.time which is alias for this

File: async_lru.py [hunk 2]
```diff
@@ -228,7 +251,10 @@ def wrapped(*fn_args, **fn_kwargs):
             wrapped.tasks.add(task)
             task.add_done_callback(wrapped.tasks.remove)
 
-            wrapped._cache[key] = task
+            if expiration_time is None:
+                wrapped._cache[key] = task
+            else:
+                wrapped._cache[key] = {"fut": task, "time": time()}
```
Comment (hellysmile) [inline async_lru.py]: what about tuple?

File: async_lru.py [hunk 3]
```diff
@@ -204,21 +206,42 @@ def wrapped(*fn_args, **fn_kwargs):
...
-                # exception here and cache_exceptions == False
-                wrapped._cache.pop(key)
+                    # exception here and cache_exceptions == False
+                    wrapped._cache.pop(key)
+
+                else:
+                    fut = wrapped._cache.get(key).get("fut")
+                    fut_time = wrapped._cache.get(key).get("time")
+
+                    if not fut.done():
```
Comment (hellysmile) [inline async_lru.py]: this and previous same block should be reused somehow, cuz it is copy-paster

File: async_lru.py (hunk 1 above)
Comment (vera1118) [reply async_lru.py]: Cool, Thanks for your reply and suggestions! I just want to double check that I can modify your code and use it in industry, right?

**Comment by Artimi** [comment]
Hi @hellysmile,
I wanted to ask whether you would accept this PR if I edited your suggestions? Because https://github.com/aio-libs/async_lru/pull/50#issuecomment-400426038 it seems that you don't want this to be in your package (and you have all rights to say that). But I can see a use case for that, especially in async world.
We are using `async_lru` as a cache for multiple coroutine call. We have a backend application that retrieves data from a database, does some computation and then sends it to frontend. It may happen that some request generates multiple calls of a coroutine with the same arguments and it's handy when I can just place an `alru_cache` decorator to save resources. However, data in the database are being updated and I need to invalidate the cache after some time. Now, easiest would be to add just `expires` keyword argument that would invalidate cache for me.
Maybe I could have a loop that invalidates the cache for me but it's now used on several places and I think it would be better to put it right into `aynsc_lru` package. 
What do you think? Will you accept the PR if I finish it?

**Comment by Artimi** [comment]
I created a new PR https://github.com/aio-libs/async_lru/pull/131

**Comment by asvetlov** [comment]
In my mind `async_lru` follows `functools.lru_cache` design, this is the very strong point.
`lru_cache` has no functionality for time-based expiration.
Why should we have it in async version of lru_cache?

I agree with @hellysmile 
If you want ttl based expiration you need another library.

**Comment by Artimi** [comment]
It would just suit me very well if `async_lru` would have the time-based expiration. The other library you are talking about will have all the features of `async_lru` plus `expires` so I will probably resolve this by having my fork of this library.
I cannot tell why `lru_cache` does not have this functionality but it certainly is one of the valid strategies: https://en.wikipedia.org/wiki/Cache_replacement_policies#Time_aware_least_recently_used_(TLRU).
Also, `alru_cache` already has more arguments / method than `lru_cache`, e.g. `cache_exceptions` argument or `invalidate` method.
If you does not want to have it the library you manage I respect it and will release my own package.

**Comment by asvetlov** [comment]
@hellysmile ?

**Comment by asvetlov** [comment]
Implemented by c46ada7

**Pull request closed without merging**